core/src/main/scala/org/apache/spark/SparkContext.scala

```diff
@@ -1410,10 +1410,12 @@ class SparkContext(config: SparkConf) extends Logging with ExecutorAllocationCli
     val scheme = new URI(schemeCorrectedPath).getScheme
     if (!Array("http", "https", "ftp").contains(scheme)) {
       val fs = hadoopPath.getFileSystem(hadoopConfiguration)
-      if (!fs.exists(hadoopPath)) {
-        throw new FileNotFoundException(s"Added file $hadoopPath does not exist.")
+      val isDir = try {
+        fs.getFileStatus(hadoopPath).isDirectory
+      } catch {
+        case f: FileNotFoundException =>
+          throw new FileNotFoundException(s"Added file $hadoopPath does not exist.").initCause(f)
```

Comment:
@rxin the exception logic is in there, hidden in the fs.exists() code; which [is just a try/catch wrapper around getFileStatus()](https://github.com/apache/hadoop/blob/trunk/hadoop-common-project/hadoop-common/src/main/java/org/apache/hadoop/fs/FileSystem.java#L1438). This patch merges back to back calls of the method (eg. exists + getFileStatus, exists+isDirectory), and also from operations like exists+open where the second call raises `FileNotFoundException` anyway. This bit of the patch is ugly because I was preserving all error messages; I'm about to submit the leaner one which passes up the FNFE directly: no try/catch